src/ray/common/ray_syncer.h

```diff
@@ -0,0 +1,294 @@
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/common/id.h"
+#include "src/ray/protobuf/ray_syncer.grpc.pb.h"
+
+namespace ray {
+namespace syncer {
+
+using ServerBidiReactor = grpc::ServerBidiReactor<ray::rpc::syncer::RaySyncMessages,
+                                                  ray::rpc::syncer::RaySyncMessages>;
+using ClientBidiReactor = grpc::ClientBidiReactor<ray::rpc::syncer::RaySyncMessages,
+                                                  ray::rpc::syncer::RaySyncMessages>;
+
+using ray::rpc::syncer::ClientMeta;
+using ray::rpc::syncer::Dummy;
+using ray::rpc::syncer::RayComponentId;
+using ray::rpc::syncer::RaySyncMessage;
+using ray::rpc::syncer::RaySyncMessages;
+using ray::rpc::syncer::RaySyncMessageType;
+using ray::rpc::syncer::ServerMeta;
+
+static constexpr size_t kComponentArraySize =
+    static_cast<size_t>(ray::rpc::syncer::RayComponentId_ARRAYSIZE);
+
+/// The interface for a reporter. Reporter is defined to be a local module which would
+/// like to let the other nodes know its status. For example, local cluster resource
+/// manager.
+struct Reporter {
+  /// Interface to get the snapshot of the component. It asks the module to take a
+  /// snapshot of the current status. Each snapshot is versioned, and it should return
+  /// std::nullopt if the version hasn't changed.
+  ///
+  /// \param current_version The version syncer module current has.
+  /// \param component_id The component id asked for.
+  virtual std::optional<RaySyncMessage> Snapshot(uint64_t current_version,
+                                                 RayComponentId component_id) const = 0;
+  virtual ~Reporter() {}
+};
+
+/// The interface for a receiver. Receiver is defined to be a module which would like
+/// to get the status of other nodes. For example, cluster resource manager.
+struct Receiver {
+  /// Interface to update a module. The module should read the `sync_message` fields and
+  /// deserialize it to update its internal status.
+  ///
+  /// \param message The message received from remote node.
+  virtual void Update(std::shared_ptr<RaySyncMessage> message) = 0;
+  virtual ~Receiver() {}
+};
+
+/// RaySyncer is an embedding service for component synchronization.
+class RaySyncer {
+ public:
+  struct ServerSyncContext;
+  /// Constructor of RaySyncer
+  ///
+  /// \param node_id The id of current node.
+  RaySyncer(const std::string &node_id);
+
+  ~RaySyncer();
+
+  /// Connect to a node. This will establish a connection between the receiver and the
+  /// current node. Right now, it'll only be called within GCS. But it can be extended.
+  /// This will make the data flow like:
+  ///    current node <- node of the stub
+  ///
+  /// \param node_id The node id to connect to
+  /// \param stub The stub for RPC operations
+  void ConnectTo(const std::string &node_id,
```

Comment:
hmm should we hide the connect-to / connect-from / node removed in a different layer?